### PR TITLE
New table column update for classifications vs manual labels

### PR DESCRIPTION
**Pull request by nsokolnikov**
New URL path, POST/[device_id]/classify

Does the same stuff as the old one, but sets a different flag in the DB to indicate that it is classified, not manually labeled.

migrate_db() included, and runs with travis to migrate the db on pull req. 

Dependencies are screwed up but deps fix is in a diff branch pull req.

### CODE REVIEW

**Comment by alexburov** [comment]
Good, I'll test it out asap.

**Comment by ianlofs** [comment]
Merge conflicts... just fix those and ill merge

**Comment by iangudger** [comment]
I have serious concerns about these changes.

**Comment by nsokolnikov** [comment]
What serious concerns @iangudger ?

File: DB/egauge_sample_log.txt [hunk 1]
```diff
@@ -21,7 +21,7 @@ User-Agent: egauge/pusher
 </data>
 </group>
 
-0
+
```
Comment (iangudger) [inline DB/egauge_sample_log.txt]: Why this change?

Comment (nsokolnikov) [inline DB/egauge_sample_log.txt]: That was a typo. Will revert

File: DB/api/migrate_db.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+        cursor = con.cursor()
+        cursor.execute("ALTER TABLE data_label ADD COLUMN classification bool NOT NULL DEFAULT TRUE;")
+    except psycopg2.DatabaseError as e:
+        print('Database error: %s' % e)
+    finally:
+        if con:
+            con.commit()
+            con.close()
+
+migrate_db()
```
Comment (iangudger) [inline DB/api/migrate_db.py]: Please add a new line at the end of this file.

**Comment by iangudger** [comment]
First, the migrate DB script does nothing and couldn't possibly work. Here is the output:
```
$ cd $GOPATH/src/github.com/seadsystem/Backend/DB/api; python ./migrate_db.py

Database error: FATAL:  role "seadapi" does not exist

Traceback (most recent call last):

  File "./migrate_db.py", line 21, in <module>

    migrate_db()

  File "./migrate_db.py", line 17, in migrate_db

    if con:

UnboundLocalError: local variable 'con' referenced before assignment
```

Doing something like that as part of the CI scripts seems like a *really* bad idea.

Second, there are no tests. The only read the checks passed was because you are synced to an old version of the CI system that doesn't check code coverage.

Third, what is the point of this change? Labeling makes sense via the API as that will be done by the web app. Classification should be done is a less static state to allow new information to improve classification of old data.

Finally, anything that mutates data really should have some sort of authentication. Letting anyone on the internet read your data is one thing, letting anyone on the internet mess up your dataset is quite another.

**Comment by nsokolnikov** [comment]
Sorry, new to databases. I was going to alter the table manually, but was told to put that in the CI script instead. 
The point is to put machine-predicted values into the DB into the same table as labels, with an extra boolean to determine whether or not it's predicted/manually input labels. 

Copied the credentials from the insert_db file - thought they would work. 

I will take out the alter from the CI script and run it manually once the pull request is complete. It only needs to be run once, ever, so it should be no problem to do.

**Comment by iangudger** [comment]
Communication with the database on the server uses Postgres peer authentication:
http://www.postgresql.org/docs/9.4/static/auth-methods.html#AUTH-PEER

We don't specify a database server or password, so it can only work locally. Besides, the CI scripts are run for each new push and pull request. A database change should only happen once. I think you misunderstood or who ever you talked to was confused. You should modify the deployment scripts which are run once when setting up a new server, not the CI scripts. Please actually test that those work after you apply your changes.

I also take strong issue with storing machine classifications in the database. We have discussed this before. Each time someone new joins I have to explain why this is a really bad idea and completely goes against good practice.

In general, you should never store something you can calculate in a database. If the base data changes, your calculated data is now stale. Store calculated data in a cache or not at all. Make it clean what is data and what isn't. Machine learning is really hard to get right and has lots of pitfalls, especially when it comes to feedback loops. This can be avoided by not allowing the data to feedback in its self by not storing results as data.

Don't do it. If you need further explanation, I will be available this weekend. Otherwise ask Ali. Maybe he remembers.

**Comment by iangudger** [comment]
Here are a couple of articles/blog posts about why you should never store calculated data in a database:
http://database-programmer.blogspot.com/2008/01/database-skills-third-normal-form-and.html
http://www.bennadel.com/blog/1190-thoughts-on-storing-duplicate-calculated-data-in-the-database.htm

**Comment by alexburov** [comment]
I agree. Ali suggested to have table with classified things on the front end side (django app). Will it be possible @nsokolnikov ?

**Comment by nsokolnikov** [comment]
Sounds good.

**Comment by aadabi** [comment]
I understand the API world and Ian is absolutely right, but They might need a one time pass to the hack  a demo ready for SEADS on Thursday. This code will be deleted on Friday.

**Comment by aadabi** [comment]
http://imgur.com/sZFz25k API is the way to go, but the question I am asking Ian L and Nikita and Alex is that can they do the right thing on time. If not,  I would go with the way they can finish the demo.

**Comment by aadabi** [comment]
The other idea that I have is :+1: 
To have a field associated with every entry about their assigned label by SEADS(RPI) . 
If the SEADS did not recognized or entered by seads it will have NL(NO Label)
Labels are numbers : 1..n which each is associated with a device.

**Comment by iangudger** [comment]
@aadabi: This does not need to be merged to allow a quick demo. When deployment scripts are added for this new code, it should be trivial to check out this branch on say a Digital Ocean droplet or AWS instance, and deploy this modified version there.

**Comment by aadabi** [comment]
Seems good, I agree with the whole notion of stateless DB. Agree we can just deploy it to AWS. And yeah it is a test branch. We need to have the main branch clean.  We can have a short discussion with @iangudger @alexburov @ianlofs  @nsokolnikov on Skype or in person about the right approaches this Saturday.

**Comment by iangudger** [comment]
@aadabi: Sounds good.

**Comment by nsokolnikov** [comment]
I assume we're discarding this branch. It's not required and we can do it some other way.
